File: Algorithms/Easy/IsomorphicStrings.go
package easy

// Time complexity: O(n)
// Space complexity: O(n)
func isIsomorphic(s string, t string) bool {
	mapST := make(map[byte]byte)
	mapTS := make(map[byte]byte)

	for i := range s {
		if _, ok := mapST[s[i]]; ok && mapST[s[i]] != t[i] {
			return false
		}

		if _, ok := mapTS[t[i]]; ok && mapTS[t[i]] != s[i] {
			return false
		}

		mapST[s[i]] = t[i]
		mapTS[t[i]] = s[i]
	}

	return true
}

File: Algorithms/Easy/205.Isomorphic-Strings/IsomorphicStrings.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        map_st, map_ts = {}, {}

        for i in range(len(s)):
            if (s[i] in map_st and map_st[s[i]] != t[i]) or (
                t[i] in map_ts and map_ts[t[i]] != s[i]
            ):
                return False

            map_st[s[i]] = t[i]
            map_ts[t[i]] = s[i]

        return True
